### PR TITLE
CHET-145: Add database migration service

### PR DESCRIPTION
**Pull request by tarka**
This add the DatabaseMigrationService and testing. As discussed IRL, this is slightly messy as the S3Uploader, etc. has slightly FS-migration-specific; the next step is to refactor-out the common areas into common components, but this is better as a 2-step process.

### CODE REVIEW

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        this.awsCredentialsProvider = awsCredentialsProvider;
+        this.regionService = regionService;
+        this.tempDirectory = tempDirectory;
+        this.endpointOverride = Optional.of(endointOverride);
+        this.setStatus(MigrationStatus.NOT_STARTED);
+    }
+
+    /**
+     * Start database dump and upload to S3 bucket. This is a blocking operation and should be started from ExecutorService
+     * or preferably from ScheduledJob. The status of the migration can be queried via
```
Comment (bpar476) [inline src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: Sad, forgotten, unfinished sentence?

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+        ConcurrentLinkedQueue<Path> uploadQueue = new ConcurrentLinkedQueue<>();
+        FileSystemMigrationProgress progress = new DefaultFilesystemMigrationProgress();
+        FileSystemMigrationErrorReport report = new DefaultFileSystemMigrationErrorReport();
+
+        S3AsyncClientBuilder builder = S3AsyncClient.builder()
+            .credentialsProvider(awsCredentialsProvider)
+            .region(Region.of(regionService.getRegion()));
+        if (endpointOverride.isPresent()) {
+            builder = builder.endpointOverride(endpointOverride.get());
```
Comment (bpar476) [inline src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: Is this mainly for the localstack tests?

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        String bucket = System.getProperty("S3_TARGET_BUCKET_NAME", "trebuchet-testing");
+        S3UploadConfig config = new S3UploadConfig(bucket, client, target.getParent());
+
+        S3Uploader uploader = new S3Uploader(config, report, progress);
+        Crawler crawler = new DirectoryStreamCrawler(report, progress);
+
+        setStatus(MigrationStatus.UPLOAD_IN_PROGRESS);
+        try {
+            crawler.crawlDirectory(target, uploadQueue);
```
Comment (bpar476) [inline src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: Is target always going to be a directory? I was under the impression that the non-parallel pg_dump outputs a single file

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        } catch (IOException e) {
+
```
Comment (bpar476) [inline src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: Should we fail the db migration if we get IOException here?

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java (hunk 2 above)
Comment (tarka) [reply src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: Yeah, this is one of the things I'll refactor. However there is a case for including it as a configuration option as a lot of AWS competitors support the S3 API.

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java (hunk 3 above)
Comment (tarka) [reply src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: For now it makes sense to use the same setup regardless; it save a bunch of special cases in the code.

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java (hunk 4 above)
Comment (tarka) [reply src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: Oops, missed that.

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java (hunk 3 above)
Comment (bpar476) [reply src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: Having a look further on I can see we're always using the directory dump method no matter what so this is probably fine 👍

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/MigrationStatus.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
+package com.atlassian.migration.datacenter.core.aws.db;
+
+import java.util.Optional;
+
+/**
+ * Copyright Atlassian: 11/03/2020
+ */
+public class MigrationStatus
```
Comment (theghostwhoforks) [inline src/main/java/com/atlassian/migration/datacenter/core/aws/db/MigrationStatus.java]: Note for later - Should use the state machine states for this instead? or do we see this as something compliementary?

File: src/main/java/com/atlassian/migration/datacenter/core/application/DatabaseConfiguration.java [hunk 6]
```diff
@@ -5,19 +5,28 @@
  */
 public class DatabaseConfiguration
 {
+    public enum DBType {
+        POSTGRESQL,
+        MYSQL,
+        SQLSERVER,
```
[the remark this answers is not in the document]
Comment (tarka) [reply src/main/java/com/atlassian/migration/datacenter/core/application/DatabaseConfiguration.java]: SQLSERVER matches the DB URI format, so it's easier to convert: https://confluence.atlassian.com/adminjiraserver/connecting-jira-applications-to-sql-server-2012-938846858.html

**Pull request merged**